taichi/program/state_flow_graph.cpp

```diff
@@ -837,6 +906,135 @@ void StateFlowGraph::verify() {
   topo_sort_nodes();
 }
 
+// TODO: make this an IR pass
+class ConstExprPropagation {
+ public:
+  static std::unordered_set<Stmt *> run(
+      Block *block,
+      const std::function<bool(Stmt *)> &is_const_seed) {
+    std::unordered_set<Stmt *> const_stmts;
+
+    auto is_const = [&](Stmt *stmt) {
+      if (is_const_seed(stmt)) {
+        return true;
+      } else {
+        return const_stmts.find(stmt) != const_stmts.end();
+      }
+    };
+
+    for (auto &s : block->statements) {
+      if (is_const(s.get())) {
+        const_stmts.insert(s.get());
+      } else if (auto binary = s->cast<BinaryOpStmt>()) {
+        if (is_const(binary->lhs) && is_const(binary->rhs)) {
+          const_stmts.insert(s.get());
+        }
+      } else if (auto unary = s->cast<UnaryOpStmt>()) {
+        if (is_const(unary->operand)) {
+          const_stmts.insert(s.get());
+        }
+      } else {
+        // TODO: ...
+      }
+    }
+
+    return const_stmts;
+  }
+};
+
+bool StateFlowGraph::demote_activation() {
+  bool modified = false;
+
+  topo_sort_nodes();
+
+  // TODO: use unordered_map
+  std::map<std::pair<IRHandle, Node *>, std::vector<Node *>> tasks;
+
+  for (int i = 1; i < (int)nodes_.size(); i++) {
+    Node *node = nodes_[i].get();
+    auto snode = node->meta->snode;
+    auto list_state = AsyncState(snode, AsyncState::Type::list);
+
+    // TODO: handle serial and range for
+    if (node->meta->type != OffloadedStmt::struct_for)
+      continue;
+
+    if (node->input_edges[list_state].size() != 1)
+      continue;
+
+    auto list_node = *node->input_edges[list_state].begin();
+    tasks[std::make_pair(node->rec.ir_handle, list_node)].push_back(node);
```

Comment:
Just to confirm, if there's such a launch sequence:

```py
@ti.kernel
def foo():
  for i in x:
    y[i] = ...

@ti.kernel
def bar():
  for i in range(...):
    if some_cond(i):
      x[i] = ...  # potentially changes x's mask/list state

foo()
bar()
foo()  # <-- cannot demote the activation of |y| now
```

Does this *not* demote activation as expected?

Reply:
I believe the things will go as expected (although not tested). The second `foo()` uses a different list of `x`, and activation demotion will only happen when the same list states are used :-)

(Btw we should set up a testing system so that cases like this can be tested...)